File: vite-vue-ts-frontend/src/modules/tasks/types/dto.ts
import type {
  PagerRequest,
  PagerResponse,
} from "../../../shared/types/dto/pager";
import type { Order } from "../../../shared/types/dto/order";

import type { TaskPriorityResponse } from "../../task-priorities/types/dto";
import type { TaskStatusResponse } from "../../task-statuses/types/dto";
import type { UserBaseResponse } from "../../users/types/dto";
import type { TimestampRange } from "../../../shared/composables/timestamps";

export type AddRequest = {
  summary: string;
  description: string | null;
  priority: {
    id: string;
  };
  status: {
    id: string;
  };
  tags: string[];
};

export type UpdateRequest = {
  id: string;
  summary: string;
  description: string | null;
  priority: {
    id: string;
  };
  status: {
    id: string;
  };
  startedAt: number | null;
  finishedAt: number | null;
  dueAt: number | null;
  tags: string[];
};

type SearchRequestFilter = {
  summary?: string;
  priorityId?: string;
  statusId?: string;
  createdAt?: TimestampRange;
  createdByUserId?: string;
};

export type SearchRequest = {
  pager: PagerRequest;
  order: Order;
  filter?: SearchRequestFilter;
};

export type TaskResponse = {
  id: string;
  projectId: string;
  slug: string;
  summary: string;
  description: string;
  priority: TaskPriorityResponse;
  status: TaskStatusResponse;
  createdAt: number;
  createdBy: UserBaseResponse;
  updatedAt?: number;
  startedAt?: number;
  finishedAt?: number;
  dueAt?: number;
  tags: string[];
  permissionsCount: number;
  attachmentsCount: number;
  notesCount: number;
  historyOperationsCount: number;
};

export type SearchResponse = {
  tasks: TaskResponse[];
  pager: PagerResponse;
};

File: golang-backend/internal/handlers/taskhandler/dto.go
package taskhandler

import (
	"github.com/aportela/doneo/internal/handlers"
	"github.com/aportela/doneo/internal/handlers/taskpriorityhandler"
	"github.com/aportela/doneo/internal/handlers/taskstatushandler"
	"github.com/aportela/doneo/internal/handlers/userhandler"
)

type taskPriority struct {
	ID string `json:"id"`
}

type taskStatus struct {
	ID string `json:"id"`
}

type addRequest struct {
	ID            string       `json:"id"`
	Summary       string       `json:"summary"`
	Description   *string      `json:"description"`
	EstimatedTime *uint64      `json:"estimatedTime"`
	Priority      taskPriority `json:"priority"`
	Status        taskStatus   `json:"status"`
	Tags          []string     `json:"tags"`
}

type updateRequest struct {
	ID            string       `json:"id"`
	Summary       string       `json:"summary"`
	Description   *string      `json:"description"`
	Priority      taskPriority `json:"priority"`
	Status        taskStatus   `json:"status"`
	StartedAt     *int64       `json:"startedAt"`
	FinishedAt    *int64       `json:"finishedAt"`
	DueAt         *int64       `json:"dueAt"`
	EstimatedTime *uint64      `json:"estimatedTime"`
	Tags          []string     `json:"tags"`
}

type filterRequest struct {
	Summary         *string                   `json:"summary"`
	PriorityID      *string                   `json:"priorityId"`
	StatusID        *string                   `json:"statusId"`
	CreatedAt       *handlers.TimestampFilter `json:"createdAt"`
	CreatedByUserID *string                   `json:"createdByUserId"`
}

type searchRequest struct {
	Pager  handlers.PagerRequest `json:"pager"`
	Order  handlers.OrderRequest `json:"order"`
	Filter *filterRequest        `json:"filter"`
}

type taskResponse struct {
	ID                     string                                   `json:"id"`
	ProjectID              string                                   `json:"projectId"`
	Slug                   string                                   `json:"slug"`
	Summary                string                                   `json:"summary"`
	Description            *string                                  `json:"description"`
	CreatedBy              userhandler.UserBaseResponse             `json:"createdBy"`
	CreatedAt              int64                                    `json:"createdAt"`
	UpdatedAt              *int64                                   `json:"updatedAt"`
	DeletedAt              *int64                                   `json:"deletedAt"`
	StartedAt              *int64                                   `json:"startedAt"`
	FinishedAt             *int64                                   `json:"finishedAt"`
	DueAt                  *int64                                   `json:"dueAt"`
	EstimatedTime          uint64                                   `json:"estimatedTime"`
	TotalSpentTime         uint64                                   `json:"totalSpentTime"`
	Priority               taskpriorityhandler.TaskPriorityResponse `json:"priority"`
	Status                 taskstatushandler.TaskStatusResponse     `json:"status"`
	Tags                   []string                                 `json:"tags"`
	TasksCount             uint16                                   `json:"tasksCount"`
	PermissionsCount       uint16                                   `json:"permissionsCount"`
	AttachmentsCount       uint16                                   `json:"attachmentsCount"`
	NotesCount             uint16                                   `json:"notesCount"`
	HistoryOperationsCount uint16                                   `json:"historyOperationsCount"`
	TimeTrackingsCount     uint16                                   `json:"timeTrackingsCount"`
}

type searchTasksResponse struct {
	Tasks []taskResponse         `json:"tasks"`
	Pager handlers.PagerResponse `json:"pager"`
}
